opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/AbstractDaoHibernate.java

```diff
@@ -259,8 +269,42 @@ public void deleteAll(final Collection<T> entities) throws DataAccessException {
 
     @Override
     public List<T> findMatching(final org.opennms.core.criteria.Criteria criteria) {
-        final HibernateCallback<List<T>> callback = buildHibernateCallback(criteria);
-        return getHibernateTemplate().execute(callback);
+        if(criteria.isMultipleAnd()){
+            return findMultiAndMatching(criteria);
+        } else {
+            final HibernateCallback<List<T>> callback = buildHibernateCallback(criteria);
+            return getHibernateTemplate().execute(callback);
+        }
+    }
+
+    private List<T> findMultiAndMatching(final org.opennms.core.criteria.Criteria criteria){
+        Set<T> allUniqueRecords = new LinkedHashSet<>();
+        Collection<Restriction> allRestrictions = criteria.getRestrictions();
+
+        //set of multiand restrictions
+        Set<Restriction> multiAndRestirctionSet = allRestrictions.stream().filter(
+                restriction -> restriction.getType().equals(Restriction.RestrictionType.MULTIAND)).collect(Collectors.toSet());
+
+        //set of non multiand restrictions
+        Set<Restriction>   nonMultiAndRestirctionSet =  Sets.difference(new HashSet<Restriction>(allRestrictions),multiAndRestirctionSet);
```

Comment:
Maybe use `Collectors.partitioningBy`?

Reply:
partitioningBy returns Map, Also, set difference is more efficient